File: python/janus/providers/factory.py
"""Provider factory for creating provider instances."""

import os
import shutil

from rich.console import Console

from ..core.types import AuthMethod, JanusConfig, ProviderConfig, ProviderType
from .anthropic import AnthropicProvider
from .base import Provider
from .claude_cli import ClaudeCliProvider
from .gemini import GeminiProvider
from .grok import GrokProvider
from .openai import OpenAIProvider

console = Console()


def _resolve_anthropic_auth_method(config: ProviderConfig) -> AuthMethod:
    """Determine which auth method to use for Anthropic."""
    # If explicitly set (not auto), use that
    if config.auth_method != AuthMethod.AUTO:
        return config.auth_method

    # Auto-detect: check for API key first
    api_key = config.api_key or os.environ.get("ANTHROPIC_API_KEY", "")

    # Check if it's a real key (not a placeholder)
    if api_key and not api_key.startswith("sk-ant-..."):
        return AuthMethod.API_KEY

    # Fall back to CLI if available
    if shutil.which("claude"):
        return AuthMethod.CLI

    # Default to API key (will fail if not configured)
    return AuthMethod.API_KEY


def create_providers(config: JanusConfig) -> list[Provider]:
    """Create provider instances based on configuration."""
    providers: list[Provider] = []

    for provider_name in config.active_providers:
        provider_config = config.providers.get(provider_name)
        if not provider_config:
            console.print(f"[yellow]Warning: Provider '{provider_name}' not configured[/yellow]")
            continue

        try:
            provider = _create_provider(provider_name, provider_config)
            if provider:
                providers.append(provider)
        except Exception as e:
            console.print(f"[red]Error creating provider '{provider_name}': {e}[/red]")

    return providers


def _create_provider(name: str, config: ProviderConfig) -> Provider | None:
    """Create a single provider instance."""
    match config.type:
        case ProviderType.ANTHROPIC:
            auth_method = _resolve_anthropic_auth_method(config)
            if auth_method == AuthMethod.CLI:
                console.print("[dim]Using Claude CLI (subscription mode)[/dim]")
                return ClaudeCliProvider(config)
            return AnthropicProvider(config)

        case ProviderType.OPENAI:
            return OpenAIProvider(config)

        case ProviderType.GEMINI:
            return GeminiProvider(config)

        case ProviderType.GROK:
            return GrokProvider(config)

        case ProviderType.OPENAI_COMPATIBLE:
            return OpenAIProvider(config, name=name.title())

        case _:
            console.print(f"[yellow]Unknown provider type: {config.type}[/yellow]")
            return None

File: typescript/src/providers/factory.ts
import { Provider } from './base.js';
import { AnthropicProvider } from './anthropic.js';
import { OpenAIProvider } from './openai.js';
import { GeminiProvider } from './gemini.js';
import { ClaudeBinaryProvider } from './claude_binary.js';
import { ConclaveConfig, ProviderConfig } from '../core/types.js';
import chalk from 'chalk';

/**
 * Determine the best auth method for Anthropic provider
 */
function resolveAnthropicAuthMethod(providerConfig: ProviderConfig): 'cli' | 'api_key' {
    const configuredMethod = providerConfig.auth_method || 'auto';

    // If explicitly set to cli or api_key, respect that
    if (configuredMethod === 'cli') {
        return 'cli';
    }

    if (configuredMethod === 'api_key') {
        return 'api_key';
    }

    // Auto mode: check for valid API key first
    const apiKey = providerConfig.api_key || process.env.ANTHROPIC_API_KEY;
    const isValidKey = apiKey &&
                       apiKey.trim() !== '' &&
                       !apiKey.includes('sk-ant-...') &&
                       !apiKey.includes('your-api-key');

    if (isValidKey) {
        return 'api_key';
    }

    // Default to CLI (subscription-based)
    return 'cli';
}

export class ProviderFactory {
    static createProviders(config: ConclaveConfig): Provider[] {
        const providers: Provider[] = [];

        // Iterate over active providers defined in the global config
        for (const providerName of config.active_providers) {
            const providerConfig = config.providers[providerName];
            if (!providerConfig) {
                console.warn(`Warning: Provider '${providerName}' configured as active but definition missing.`);
                continue;
            }

            try {
                switch (providerConfig.type) {
                    case 'openai':
                        providers.push(new OpenAIProvider(providerConfig));
                        break;

                    case 'anthropic': {
                        const authMethod = resolveAnthropicAuthMethod(providerConfig);

                        if (authMethod === 'api_key') {
                            providers.push(new AnthropicProvider(providerConfig));
                        } else {
                            console.log(chalk.blue('ℹ Using Claude CLI (subscription mode)'));
                            providers.push(new ClaudeBinaryProvider(providerConfig));
                        }
                        break;
                    }

                    case 'gemini':
                        providers.push(new GeminiProvider(providerConfig));
                        break;

                    case 'openai_compatible':
                    case 'grok':
                        // Re-use OpenAI provider for generic endpoints (Grok, Ollama, etc)
                        providers.push(new OpenAIProvider(providerConfig));
                        break;

                    default:
                        console.warn(`Unknown provider type: ${(providerConfig as any).type}`);
                }
            } catch (error) {
                console.error(`Error initializing provider ${providerName}:`, error);
            }
        }

        return providers;
    }
}
